### PR TITLE
chats: added openAi chat generation as LLM, fixes #43

### LINKED ISSUE #43: Introduce `chatModels`
I already started with openai but have the need to discuss how we want to map langchain to langchaingo in a few instances. I'll make a PR soon.

### PR DESCRIPTION
**Pull request by ChristianSch**
### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [ ] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [ ] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [ ] Describes source of new concepts.
- [ ] References existing implementations as appropriate.
- [ ] Contains test coverage for new functions.
- [ ] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.
- [ ] (remove and do not use this row unless you are a trusted contributor) I am an existing
  contributor.

### CODE REVIEW

File: chatmodels/openai/internal/openaiclient/chat.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+)
+
+// ChatPayloadMessage is a message in the format expected by the OpenAI API.
+type ChatPayloadMessage struct {
```
Comment (ChristianSch) [inline chatmodels/openai/internal/openaiclient/chat.go:11]: as already said in the comments in https://github.com/tmc/langchaingo/issues/43 this format is currently not compatible with the LLM interface but is required by the openAI API.

**Comment by ChristianSch** [comment]
This, as is, not compatible with the internal API of this library. That's why this is a draft and doesn't include tests or anything. I would like to discuss how to proceed. 

To test the implementation, use the following main.go file:
https://gist.github.com/ChristianSch/20251608147883cd5c6752988f16eb8e

File: chatmodels/openai/openaichatllm.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	return []*llms.Generation{
+		{
+			Text:           result.Text,
+			GenerationInfo: map[string]any{},
+		},
+	}, nil
+}
+
+func New() (*LLM, error) {
```
Comment (ChristianSch) [inline chatmodels/openai/openaichatllm.go:59]: right now none of the LLMs are properly configurable by this contract. what I would like to do is something like this:

```go
type OpenAIChatConfig struct {
    Model       string
    // note the absence of "Messages" here, is provided in Call/Generate
    Temperature float64  `json:"temperature"`
    N           int      `json:"n"`
    MaxTokens   int      `json:"max_tokens"`
    Stop        []string `json:"stop,omitempty"` // could be moved to Call/Generate as well?
}

func New(config OpenAIChatConfig) (*LLM, error) {
    TODO()
}
```

Comment (ChristianSch) [reply chatmodels/openai/openaichatllm.go:59]: this would break with the rest of the LLMs but currently makes this lib not really usable imho

Comment (tmc) [reply chatmodels/openai/openaichatllm.go:59]: Let's do variadic/functional Options arguments to pass these (and have some reasonable defaults).

like

```go
func New(options ...Option) (*LLM, error) {
```

**Review by tmc — COMMENTED** [review]
I don't think we should make a separate client here, instead we should move/refactor/extend the openai client used in the llm tree.

**Comment by tmc** [comment]
I'm going to accept 56 and we can refactor this as needed.

**Comment by tmc** [comment]
This is solved in other work/ merged PRs.

**Pull request closed without merging**